conn_browser.go

```diff
@@ -0,0 +1,311 @@
+		c.closeHandler.Release()
+		c.closeHandler = nil
+	}
+	if c.errorHandler != nil {
+		c.Call("removeEventListener", "error", *c.errorHandler)
+		c.errorHandler.Release()
+		c.errorHandler = nil
+	}
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	return c.localAddr
+}
+
+func getRemoteAddr(val js.Value) net.Addr {
+	rawURL := val.Get("url").String()
+	withoutPrefix := strings.TrimPrefix(rawURL, "ws://")
+	withoutSuffix := strings.TrimSuffix(withoutPrefix, "/")
+	return NewAddr(withoutSuffix)
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	return c.remoteAddr
+}
+
+// TODO: Return os.ErrNoDeadline. For now we return nil because multiplexers
+// don't handle the error correctly.
+func (c *Conn) SetDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetReadDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetWriteDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) setUpHandlers() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if c.messageHandler != nil {
+		// Message handlers already created. Nothing to do.
+		return
+	}
+	messageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		arrayBuffer := args[0].Get("data")
+		data := arrayBufferToBytes(arrayBuffer)
+		c.currDataMut.Lock()
+		if _, err := c.currData.Write(data); err != nil {
+			c.currDataMut.Unlock()
+			return err
+		}
+		c.currDataMut.Unlock()
+
+		// Non-blocking signal
+		select {
+		case c.dataSignal <- struct{}{}:
+		default:
+		}
+
+		return nil
+	})
+	c.messageHandler = &messageHandler
+	c.Call("addEventListener", "message", messageHandler)
+
+	closeHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			c.signalClose()
+			c.mut.Lock()
+			// Store the error in c.firstErr. It will be returned by Read later on.
+			c.firstErr = errorEventToError(args[0])
+			c.mut.Unlock()
+			c.releaseHandlers()
+		}()
+		return nil
+	})
+	c.closeHandler = &closeHandler
+	c.Call("addEventListener", "close", closeHandler)
+
+	errorHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		// Unfortunately, the "error" event doesn't appear to give us any useful
+		// information. All we can do is close the connection.
+		c.Close()
+		return nil
+	})
+	c.errorHandler = &errorHandler
+	c.Call("addEventListener", "error", errorHandler)
+}
+
+func (c *Conn) waitForOpen() error {
+	openSignal := make(chan struct{})
+	handler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		close(openSignal)
+		return nil
+	})
+	defer c.Call("removeEventListener", "open", handler)
+	defer handler.Release()
+	c.Call("addEventListener", "open", handler)
+	select {
+	case <-openSignal:
+		return nil
+	case <-c.closeSignal:
+		// c.closeSignal means there was an error when trying to open the
+		// connection.
+		return c.firstErr
+	}
+}
+
+// arrayBufferToBytes converts a JavaScript ArrayBuffer to a slice of bytes.
+func arrayBufferToBytes(buffer js.Value) []byte {
```

Comment:
So... go 1.13 is out. Want to go ahead and use `CopyBytesToGo` and `CopyBytesToJs`?

Reply:
My personal preference would be to release a version that works with Go 1.12 and _then_ include this optimization. There are some speed bumps with upgrading to Go 1.13 so some projects (0x Mesh included) might not have updated yet.

Reply:
SGTM. Note: libp2p is likely to update to 1.13 before too long as we need it for ed25519+TLS.